File: framework/languages/node/packages/stream-connector/src/Runtime/ZlinkStreamRuntimeMetrics.ts
import type { RequiredZlinkStreamConnectorOptions } from '../Contracts';

export class ZlinkStreamRuntimeMetrics {
  private readonly reconnects;
  private readonly handshakeDuration;
  private readonly handshakeFailures;
  private readonly inboundBytes;
  private readonly outboundBytes;

  constructor(private readonly options: RequiredZlinkStreamConnectorOptions) {
    const meter = options.meterProvider?.getMeter('zlink.framework');
    this.reconnects = meter?.createCounter('zlink.stream.reconnects', { unit: '{event}' });
    this.handshakeDuration = meter?.createHistogram('zlink.stream.handshake.duration', { unit: 's' });
    this.handshakeFailures = meter?.createCounter('zlink.stream.handshake.failures', { unit: '{failure}' });
    this.inboundBytes = meter?.createCounter('zlink.stream.inbound.bytes', { unit: 'By' });
    this.outboundBytes = meter?.createCounter('zlink.stream.outbound.bytes', { unit: 'By' });
  }

  reconnect(): void {
    this.safe(() => this.reconnects?.add(1, { transport: this.transportLabel() }));
  }

  handshakeCompleted(startedAt: number): void {
    const seconds = (performance.now() - startedAt) / 1000;
    this.safe(() => this.handshakeDuration?.record(seconds, { transport: this.transportLabel() }));
  }

  handshakeFailed(error: unknown): void {
    const reason = error instanceof Error && (error.name === 'AbortError' || error.name === 'TimeoutError')
      ? 'canceled'
      : 'transport_error';
    this.safe(() => this.handshakeFailures?.add(1, { transport: this.transportLabel(), reason }));
  }

  inbound(byteCount: number): void {
    this.safe(() => this.inboundBytes?.add(byteCount, { transport: this.transportLabel() }));
  }

  outbound(byteCount: number): void {
    this.safe(() => this.outboundBytes?.add(byteCount, { transport: this.transportLabel() }));
  }

  private transportLabel(): 'ws' | 'wss' {
    return this.options.endpoint.startsWith('wss:') ? 'wss' : 'ws';
  }

  private safe(record: () => void): void {
    try {
      record();
    } catch {
      // Metrics listeners are external observation code and cannot change connector flow.
    }
  }
}

File: framework/languages/dotnet/src/Systems.Zlink.Stream.Connector/Runtime/ZlinkStreamRuntimeMetrics.cs
using System.Diagnostics.Metrics;

namespace Systems.Zlink.Stream.Connector.Runtime;

internal static class ZlinkStreamRuntimeMetrics
{
    // Every record method contains the listener boundary. Metrics callbacks
    // are external observation code and must never escape into connector flow.
    internal const string MeterName = "zlink.framework.stream_connector";

    private static readonly Meter Meter = new(MeterName);

    private static readonly Counter<long> Reconnects =
        Meter.CreateCounter<long>("zlink.stream.reconnects", "{reconnect}");

    internal static void RecordReconnect(
        string transport,
        string outcome,
        string reason)
    {
        try
        {
            if (!Reconnects.Enabled) return;
            Reconnects.Add(
                1,
                new KeyValuePair<string, object?>("transport", transport),
                new KeyValuePair<string, object?>("outcome", outcome),
                new KeyValuePair<string, object?>("reason", reason));
        }
        catch
        {
        }
    }

    internal static string TransportLabel(ZlinkStreamConnectorOptions options)
    {
        return options.Endpoint.Scheme switch
        {
            "tcp" => "tcp",
            "tls" => "tls",
            "ws" => "ws",
            "wss" => "wss",
            _ => "unknown"
        };
    }
}
